### PR TITLE
[Elastic Agent] Enable log shipping of endpoint-security by Elastic Agent

### LINKED ISSUE #21796: [Elastic Agent] Allow the agent to collect and send the log of endpoint.
We need to add support for collecting endpoint logs and send them to the following data_stream: `logs-elastic_agent.endpoint-{namespace}` with no configured namespace it will be `logs-elastic_agent.endpoint-default`. 

@ferullo Are you writing JSON formatted logs or using only plaintext?

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Starts shipping the logs of endpoint-security to elasticsearch.

This also refactors the code some to pass the `program.Spec` around instead of needing to keep checking for it from the SupportedMap. This is how the code path determines the path for the endpoint-security log paths.

## Why is it important?

So all log information from an Elastic Agent running endpoint-security can be observed.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [X] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Build the Elastic Agent with endpoint-security manually placed in the download folder. Enable Endpoint Security in Fleet and notice that all the Endpoint Security logs show up with `event.dataset: elastic_agent.endpoint-security`

## Related issues

- Closes #21796

### CODE REVIEW

**Comment by ph** [comment]
@blakerouse do we have [everything we need in the data send](https://github.com/elastic/beats/issues/21796) from endpoint to make sure we can display the log of the endpoint in the logstream UI made by @jen-huang. ?

**Comment by blakerouse** [comment]
@ph Yes it included the agent ID in the events so the LogStream component can show these logs.

File: x-pack/elastic-agent/pkg/agent/application/upgrade/upgrade.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/application/upgrade/upgrade.go:15]: how this don't bother our formatter

File: x-pack/elastic-agent/pkg/artifact/install/tar/tar_installer.go [hunk 2]
```diff
@@ -14,6 +14,8 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
+
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/artifact/install/tar/tar_installer.go:18]: unnecessary newline

File: x-pack/elastic-agent/pkg/agent/operation/operator.go [hunk 3]
```diff
@@ -256,7 +256,7 @@ func (o *Operator) getApp(p Descriptor) (Application, error) {
 		return a, nil
 	}
 
-	specifier, ok := p.(app.Specifier)
+	desc, ok := p.(*app.Descriptor)
 	if !ok {
 		return nil, fmt.Errorf("descriptor is not an app.Specifier")
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/operation/operator.go:261]: can you align error message as well?

File: x-pack/elastic-agent/pkg/core/plugin/process/app.go [hunk 4]
```diff
@@ -12,6 +12,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/core/plugin/process/app.go:15]: i guess goland just likes newlines

**Review by michalpristas — APPROVED** [review]
I like the change, probably will make future changes simpler.
i left a comments but they are related to segmented imports. nothing critical

File: x-pack/elastic-agent/pkg/core/plugin/process/app.go (hunk 4 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/core/plugin/process/app.go]: yes it does, I spend to much time trying to re-order imports

File: x-pack/elastic-agent/pkg/agent/operation/operator.go (hunk 3 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/operation/operator.go]: fixed

**Comment by ph** [comment]
@blakerouse Can add details on the following for testing that change:

AC:

- When endpoint integration is installed and runnings.
- Endpoint will write logs at :
  darwin: "/Library/Elastic/Endpoint/state/log/endpoint-*.log"
  linux: "/opt/Elastic/Endpoint/state/log/endpoint-*.log"
  windows: "C:\\Program Files\\Elastic\\Endpoint\\state\\log\\endpoint-*.log"
- Endploint Log events will be sent to the datastream **XXXX**

Fyi @EricDavisX

**Comment by blakerouse** [comment]
Should really need to worry about the log location, well unless it doesn't work ;-)

The events will have show up in the Logs UI. Easiest to filter by `event.dataset: elastic_agent.endpoint-security`

**Comment by ph** [comment]
@blakerouse but there are sent to a distinct datastream right?

**Comment by blakerouse** [comment]
@ph Yes should be similar to dataset name `logs-elastic_agent-endpoint-security-default`.

**Comment by EricDavisX** [comment]
The PR and the issue are closed (which is ok!) so I've logged a test tracking issue for any remaining discussion questions about test coverage. https://github.com/elastic/beats/issues/22549

**Comment by ph** [comment]
@blakerouse We can't use `logs-elastic_agent-endpoint-security-default` this will cause issue when we are extracting dataset and namespace. I will create an issue. I think the other datastreams too are problematic.

**Comment by ph** [comment]
Created this issue as a followup https://github.com/elastic/beats/issues/22551

**Pull request merged**